### PR TITLE
Enforce DB transaction's order to prevent deadlocks

### PR DESCRIPTION
**Pull request by pasqu4le**
fixes: #2587 

## Motivation

Due to the ShareLock mechanism used by PostgreSQL when there are several DB transactions are acting on multiple rows of the same table, it's possible to incur in a deadlock and so into an error.

## Changelog

This PR enforces a consistent order of ShareLock acquisition in all the relevant transaction.
It also adds a clear documentation that should help in the future.

### Bug Fixes

This is the list of functions that had to be modified and for what reason: 

Updates that were performed without ordering:
- Explorer.Chain.find_and_update_replaced_transactions
- Explorer.Chain.update_replaced_transactions
- Explorer.Chain.Import.Runner.Addresses.update_transactions
- Explorer.Chain.Import.Runner.Blocks.update_internal_transaction_block_number
- Explorer.Chain.Import.Runner.Blocks.fork_transactions
- Explorer.Chain.Import.Runner.Blocks.lose_consensus
- Explorer.Chain.Import.Runner.Blocks.lose_invalid_neighbour_consensus
- Explorer.Chain.Import.Runner.Blocks.update_block_second_degree_relations
- Indexer.Temporary.BlocksTransactionsMismatch.run_blocks

Updates that were incorrectly ordered (e.g. postgres does not respect the order of a list to match)
- Explorer.Chain.Import.Runner.InternalTransactions.update_transactions
- Explorer.Chain.Import.Runner.InternalTransactionsIndexedAtBlocks.update_blocks
- Explorer.Chain.Import.Runner.Transactions.discard_blocks_for_recollated_transactions
- Explorer.Chain.Import.Runner.Blocks.derive_transaction_forks

Deletes without order
- Explorer.ChainSpec.Parity.Importer.import_emission_rewards
- Explorer.Chain.Import.Runner.Blocks.delete_rewards

Ordering that did not exist at all (and the functions that now use them):
- StakingPool - staking_address_hash
  - Explorer.Chain.Import.Runner.StakingPools.mark_as_deleted
  - Explorer.Chain.Import.Runner.StakingPools.insert
- EmissionReward - block_range
  - Explorer.ChainSpec.Parity.Importer.import_emission_rewards
- StakingPoolsDelegators delegator_address_hash, pool_address_hash
  - Explorer.Chain.Import.Runner.StakingPoolsDelegators.insert
- MarketHistory - date
  - Explorer.Market.bulk_insert_history
- Address.Name - address_hash
  - Explorer.Validator.MetadataImporter.import_metadata
- ContractMethod - {identifier, abi}
  - Explorer.Chain.ContractMethod.upsert_from_abi

SQL.query calls where refactored for clarity and ordering was added where necessary:
- Explorer.Chain.Import.Runner.Blocks.derive_transaction_forks
- Explorer.Chain.Import.Runner.Blocks.derive_address_current_token_balances
- Explorer.Chain.Import.Runner.Tokens.update_holder_counts_with_deltas

#### Modified ordering

These are orderings that were found to be inconsistent. All the functions that use them (including those corrected in the list above), have been changed to use the new ones.
The modified ones are:
- for `Block`, from `number, hash` to `hash`
- for `Transaction.Fork`,  from `uncle_hash, hash` to `uncle_hash, index`
- for `Address.Name`, from `address_hash` to `address_hash, name`

## Checklist for your PR

  - [x] I added an entry to `CHANGELOG.md` with this PR
  - [x] If I added new functionality, I added tests covering it.
  - [x] If I fixed a bug, I added a regression test to prevent the bug from silently reappearing again.
  - [x] I checked whether I should update the docs and did so if necessary
  - [x] If I added/changed/removed ENV var, I should update the list of env vars in https://github.com/poanetwork/blockscout/blob/master/docs/env-variables.md to reflect changes in the table here https://poanetwork.github.io/blockscout/#/env-variables?id=blockscout-env-variables. I've set `master` in the `Version` column.
  - [x] If I add new indices into DB, I checked, that they don't redundant with PGHero or other tools

### CODE REVIEW

File: apps/explorer/lib/explorer/chain.ex [hunk 1]
```diff
@@ -2865,7 +2872,11 @@ defmodule Explorer.Chain do
         where: pending.hash in ^hashes and is_nil(pending.block_hash)
       )
 
-    Repo.update_all(transactions_to_update, [set: [error: "dropped/replaced", status: :error]], timeout: timeout)
+    Repo.update_all(
```
Comment (ayrat555) [inline apps/explorer/lib/explorer/chain.ex]: this impression expression is not changed. Some text editors when using `mix format` in umbrella applications don't respect the project's formatter config. I had the same problem with `emacs`

File: docs/sharelocks.md [hunk 2]
```diff
@@ -0,0 +1,88 @@
+<!--sharelocks.md -->
```
Comment (ayrat555) [inline docs/sharelocks.md]: :+1: :+1: :+1:

File: apps/explorer/lib/explorer/chain.ex [hunk 3]
```diff
@@ -2846,14 +2847,20 @@ defmodule Explorer.Chain do
...
     query =
-      Enum.reduce(transactions, Transaction, fn %{hash: hash, nonce: nonce, from_address_hash: from_address_hash},
-                                                query ->
-        from(t in query,
-          or_where:
-            t.nonce == ^nonce and t.from_address_hash == ^from_address_hash and t.hash != ^hash and
-              not is_nil(t.block_number)
-        )
-      end)
+      transactions
```
[the remark this answers is not in the document]
Comment (pasqu4le) [reply apps/explorer/lib/explorer/chain.ex]: There is a comment that might be the cause of confusion, but following the original call to `Enum.reduce` there is more piping.

File: apps/explorer/lib/explorer/chain.ex (hunk 1 above)
Comment (pasqu4le) [reply apps/explorer/lib/explorer/chain.ex]: You are right, I'll put it back as it was

**Pull request merged**